crates/autopilot/src/domain/competition/participant.rs

```diff
@@ -1,62 +1,88 @@
-use {
-    super::{Score, Solution},
-    crate::infra,
-    std::sync::Arc,
-};
+use {super::Score, crate::infra, std::sync::Arc};
 
 #[derive(Clone)]
 pub struct Participant<State = Ranked> {
-    solution: Solution,
+    solution: super::Solution,
     driver: Arc<infra::Driver>,
     state: State,
 }
 
 #[derive(Clone)]
-pub struct Unranked;
-pub enum Ranked {
+pub struct Unscored;
+
+#[derive(Clone)]
+pub struct Scored {
+    pub(super) score: Score,
+}
+
+#[derive(Clone)]
+pub struct Ranked {
+    pub(super) rank_type: RankType,
+    pub(super) score: Score,
+}
+
+#[derive(Clone)]
+pub enum RankType {
     Winner,
     NonWinner,
     FilteredOut,
 }
 
 impl<T> Participant<T> {
-    pub fn solution(&self) -> &Solution {
+    pub fn solution(&self) -> &super::Solution {
         &self.solution
     }
 
-    pub fn set_score(&mut self, score: Score) {
-        self.solution.score = Some(score);
-    }
-
     pub fn driver(&self) -> &Arc<infra::Driver> {
         &self.driver
     }
 }
 
-impl Participant<Unranked> {
-    pub fn new(solution: Solution, driver: Arc<infra::Driver>) -> Self {
+impl Participant<Unscored> {
+    pub fn new(solution: super::Solution, driver: Arc<infra::Driver>) -> Self {
         Self {
             solution,
             driver,
-            state: Unranked,
+            state: Unscored,
         }
     }
 
-    pub fn rank(self, rank: Ranked) -> Participant<Ranked> {
-        Participant::<Ranked> {
-            state: rank,
+    pub fn with_score(self, score: Score) -> Participant<Scored> {
+        Participant {
             solution: self.solution,
             driver: self.driver,
+            state: Scored { score },
+        }
+    }
+}
+
+impl Participant<Scored> {
+    pub fn score(&self) -> Score {
+        self.state.score
+    }
+
+    pub fn rank(self, rank_type: RankType) -> Participant<Ranked> {
```

Comment:
We can keep consistency with the other PR by renaming this to `with_rank`, following the convention that `with_...` methods return the new state and plain `score` or `rank` are just accessors.